src/ctapipe/tools/optimize_event_selection.py

```diff
@@ -0,0 +1,234 @@
+        directory_ok=False,
+        help="Output file storing optimization result",
+    ).tag(config=True)
+
+    obs_time = AstroQuantity(
+        default_value=u.Quantity(50, u.hour),
+        physical_type=u.physical.time,
+        help=(
+            "Observation time in the form ``<value> <unit>``."
+            " This is used for flux normalization when calculating sensitivities."
+        ),
+    ).tag(config=True)
+
+    optimization_algorithm = traits.ComponentName(
+        CutOptimizerBase,
+        default_value="PointSourceSensitivityOptimizer",
+        help="The cut optimization algorithm to be used.",
+    ).tag(config=True)
+
+    aliases = {
+        "gamma-file": "EventSelectionOptimizer.gamma_file",
+        "proton-file": "EventSelectionOptimizer.proton_file",
+        "electron-file": "EventSelectionOptimizer.electron_file",
+        "output": "EventSelectionOptimizer.output_path",
+        "chunk_size": "EventSelectionOptimizer.chunk_size",
+    }
+
+    classes = [EventLoader] + classes_with_traits(CutOptimizerBase)
+
+    def setup(self):
+        """
+        Initialize components from config.
+        """
+        self.optimizer = CutOptimizerBase.from_name(
+            self.optimization_algorithm, parent=self
+        )
+        self.particles = [
+            EventLoader(
+                parent=self,
+                kind="gammas",
+                file=self.gamma_file,
+                target_spectrum=self.gamma_target_spectrum,
+            )
+        ]
+        if not isinstance(self.optimizer, PercentileCuts):
+            if not self.proton_file or (
+                self.proton_file and not self.proton_file.exists()
+            ):
+                raise ValueError(
+                    "Need a proton file for cut optimization "
+                    f"using {self.optimization_algorithm}."
+                )
+
+            self.particles.append(
+                EventLoader(
+                    parent=self,
+                    kind="protons",
+                    file=self.proton_file,
+                    target_spectrum=self.proton_target_spectrum,
+                )
+            )
+            if self.electron_file and self.electron_file.exists():
+                self.particles.append(
+                    EventLoader(
+                        parent=self,
+                        kind="electrons",
+                        file=self.electron_file,
+                        target_spectrum=self.electron_target_spectrum,
+                    )
+                )
+            else:
+                self.log.warning("Optimizing cuts without electron file.")
+
+    def start(self):
+        """
+        Load events and optimize g/h (and theta) cuts.
+        """
+        reduced_events = dict()
+        for sel in self.particles:
+            evs, cnt, meta = sel.load_preselected_events(self.chunk_size, self.obs_time)
+            if isinstance(self.optimizer, PointSourceSensitivityOptimizer):
```

Comment:
It seems kind of an anti-pattern to define an abstract API, but then to do things specifically for one particular implementation. This essentially means that the interface is missing parts.

E.g. here there could be a property `self.optimizer.needs_event_weights`  which would then be a general property any `Optimizer` could set